### DISCUSSION TITLE
azd pipeline config got error 401

### DISCUSSION THREAD

**Discussion by thanhrieker** [Azure/azure-dev#4240] [Q&A]
Good day,

Using "azd pipeline config" before running well as one day the service principal with the password expired (re-generated new password). Also PAT expired -> regenerate too.

After that I got error: "ERROR: Request returned status: 401 Unauthorized" I have no idea how to fix this issue.

Can someone please help me?

The debug file attached.
[azd_debug.txt](https://github.com/user-attachments/files/16724931/azd_debug.txt)

Thank you
Thanh

**Comment by vhvb1989** [comment]
What azd version are you using?  (run azd version)

Can you try setting a new service principal by running `azd pipeline config --principal-name sp-name --auth-type federated` .  Azd should then create a new SP and a federated credential for it (instead of client secret) if you are using the latest azd version.

**Reply by rajeshkamal5050** [reply]
Attached debug log says 1.9.6 @vhvb1989 

2024/08/23 13:13:50 main.go:56: azd version: 1.9.6 (commit 0b1a8ad350177078312141f4465ad72205d7c077)

**Reply by thanhrieker** [reply]
I use lasted version: azd version 1.9.6 (commit 0b1a8ad350177078312141f4465ad72205d7c077)

I try with your command but I got error:

ERROR: pre-config check error from Azure DevOps provider: Azure DevOps does not support federated authentication. To explicitly use client credentials set the `--auth-type client-credentials` flag. pipeline authentication configuration is not supported.

Thanks
Thanh

**Reply by vhvb1989** [reply]
oh!  looks like you are using Terraform, that's why fed credential is not supported for your case.

Did you try `azd pipeline config --principal-name sp-name`

Just to see if you are able to create a new SP with client-secret or not

**Reply by thanhrieker** [reply]
I did but the same error message...

**Reply by thanhrieker** [reply]
Good day,

I updated to lasted version
` 2024/09/18 13:38:29 main.go:56: azd version: 1.10.1 (commit 31409a33266fb4a5fdbb644bc83988e725d6c7c9)` and try again then get different error message:

```
2024/09/18 13:38:46 main.go:50: Retry: response 200
2024/09/18 13:38:46 main.go:50: Retry: exit due to non-retriable status code
  (✓) Done: Configuring client credentials for service principal

ERROR: creating new azdo client: Request returned status: 401 Unauthorized
```

Best
Thanh

**Reply by vhvb1989** [reply]
The `401 Unauthorized` means the PAT you are using might be expired or has not enough permissions to access your Org

**Reply by rajeshkamal5050** [reply]
@thanhrieker are you still blocked? any updates on this from your side?

**Reply by thanhrieker** [reply]
> The `401 Unauthorized` means the PAT you are using might be expired or has not enough permissions to access your Org

Generate new PAT the same. I follow up the article:

[https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/configure-devops-pipeline?tabs=azdo](https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/configure-devops-pipeline?tabs=azdo)

I got no error like this before.

> @thanhrieker are you still blocked? any updates on this from your side?

Yes, still blocked
[azd_debug(2).txt](https://github.com/user-attachments/files/17110898/azd_debug.2.txt)
.

Best
Thanh

**Reply by vhvb1989** [reply]
Are you setting `AZURE_DEVOPS_EXT_PAT` env var?  or how are you setting up the PAT ?

Can you share the list of roles/access you granted for the PAT?

**Reply by thanhrieker** [reply]
Good day,

I setting AZURE_DEVOPS_EXT_PAT. And I using command export AZURE_DEVOPS_EXT_PAT=...

Now I using command:

`azd env set AZURE_DEVOPS_EXT_PAT $value`

Then it is working now.

Thanks you for your help
Thanh

**Reply by vhvb1989** [reply]
It should work using `export AZURE_DEVOPS_EXT_PAT=....` the same as using `azd env set ...`  
It is actually recommended to use the first one and keep your PAT on system ENV instead of azd .env.

Sometimes `pat` has `=` symbol or other chars which can be recognized by your terminal and change the actual value that is saved to the env var.
After running `export AZURE_DEVOPS_EXT_PAT=...` , you can try `echo $AZURE_DEVOPS_EXT_PAT` and validate.
Ideally, you can run `export AZURE_DEVOPS_EXT_PAT="yout-pat-here"` .  By adding the `"`, you can avoid parsing issues from the terminal.

**Reply by thanhrieker** [reply]
I did the same with your idea by using export AZURE_DEVOPS_EXT_PAT="your-pat-here" but it is not working.

Compare with echo $AZURE_DEVOPS_EXT_PAT is the same with azd env set. But yesterday it is not working.

Today I try: azd env set AZURE_DEVOPS_EXT_PAT $value

Then it is working without any error.

Best
Thanh

**Reply by vhvb1989** [reply]
Can you share what OS you are using @thanhrieker ?  Is it Mac or some Linux distro? (or Windows+WSL?)
Is it x64 or ARM?
And what terminal are you using?  (VSCode + cmd?   sh?  bash?  )

thank you.

**Reply by thanhrieker** [reply]
I'm using Mac Sonoma 14.6.1, M1 CPU

I'm using VSCode + terminal.

Thanks
Thanh

**Reply by vhvb1989** [reply]
@weikanglim do you have an M1 Mac to try this?   I am wondering if there might be a bug when handling the PAT coming from the system env.  But I couldn't repro on Win/Lin

**Comment by webstean** [comment]
Please use OIDC Federation (OpenID Connect) for better security, that way you require no secrets.
[Setting up Terraform Azure provider to use OIDC Federation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/service_principal_oidc)
[Setting up Terraform Entra ID provider to use OIDC Federation](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/guides/service_principal_oidc)

**Reply by thanhrieker** [reply]
> Please use OIDC Federation (OpenID Connect) for better security, that way you require no secrets. [Setting up Terraform Azure provider to use OIDC Federation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/service_principal_oidc) [Setting up Terraform Entra ID provider to use OIDC Federation](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/guides/service_principal_oidc)

Thanks you for your information.

From last try I got this error:

`ERROR: pre-config check error from Azure DevOps provider: Azure DevOps does not support federated authentication. To explicitly use client credentials set the --auth-type client-credentials flag. pipeline authentication configuration is not supported.`

I will retry this time again.

Get back to you.

Best
Thanh
